pinot-core/src/main/java/org/apache/pinot/core/operator/timeseries/TimeSeriesAggregationOperator.java

```diff
@@ -0,0 +1,251 @@
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.lang3.NotImplementedException;
+import org.apache.pinot.common.request.context.ExpressionContext;
+import org.apache.pinot.core.common.BlockValSet;
+import org.apache.pinot.core.common.Operator;
+import org.apache.pinot.core.operator.BaseOperator;
+import org.apache.pinot.core.operator.BaseProjectOperator;
+import org.apache.pinot.core.operator.ExecutionStatistics;
+import org.apache.pinot.core.operator.blocks.ValueBlock;
+import org.apache.pinot.core.operator.blocks.results.TimeSeriesResultsBlock;
+import org.apache.pinot.tsdb.spi.AggInfo;
+import org.apache.pinot.tsdb.spi.TimeBuckets;
+import org.apache.pinot.tsdb.spi.series.BaseTimeSeriesBuilder;
+import org.apache.pinot.tsdb.spi.series.TimeSeries;
+import org.apache.pinot.tsdb.spi.series.TimeSeriesBlock;
+import org.apache.pinot.tsdb.spi.series.TimeSeriesBuilderFactory;
+
+
+/**
+ * Segment level operator which converts data in relational model to a time-series model, by aggregating on a
+ * configured aggregation expression.
+ */
+public class TimeSeriesAggregationOperator extends BaseOperator<TimeSeriesResultsBlock> {
+  private static final String EXPLAIN_NAME = "TIME_SERIES_AGGREGATION";
+  private final String _timeColumn;
+  private final TimeUnit _storedTimeUnit;
+  private final Long _timeOffset;
+  private final AggInfo _aggInfo;
+  private final ExpressionContext _valueExpression;
+  private final List<String> _groupByExpressions;
+  private final BaseProjectOperator<? extends ValueBlock> _projectOperator;
+  private final TimeBuckets _timeBuckets;
+  private final TimeSeriesBuilderFactory _seriesBuilderFactory;
+
+  public TimeSeriesAggregationOperator(
+      String timeColumn,
+      TimeUnit timeUnit,
+      Long timeOffset,
+      AggInfo aggInfo,
+      ExpressionContext valueExpression,
+      List<String> groupByExpressions,
+      TimeBuckets timeBuckets,
+      BaseProjectOperator<? extends ValueBlock> projectOperator,
+      TimeSeriesBuilderFactory seriesBuilderFactory) {
+    _timeColumn = timeColumn;
+    _storedTimeUnit = timeUnit;
+    _timeOffset = timeOffset;
+    _aggInfo = aggInfo;
+    _valueExpression = valueExpression;
+    _groupByExpressions = groupByExpressions;
+    _projectOperator = projectOperator;
+    _timeBuckets = timeBuckets;
+    _seriesBuilderFactory = seriesBuilderFactory;
+  }
+
+  @Override
+  protected TimeSeriesResultsBlock getNextBlock() {
+    ValueBlock transformBlock = _projectOperator.nextBlock();
+    BlockValSet blockValSet = transformBlock.getBlockValueSet(_timeColumn);
+    long[] timeValues = blockValSet.getLongValuesSV();
+    if (_timeOffset != null && _timeOffset != 0L) {
+      timeValues = applyTimeshift(_timeOffset, timeValues);
+    }
+    int[] timeValueIndexes = getTimeValueIndex(timeValues, _storedTimeUnit);
+    Object[][] tagValues = new Object[_groupByExpressions.size()][];
```

Comment:
Yup this will work. Group by can have arbitrary expressions. Though I need to rename `ScanFilterAndProjectPlanNode` to something like `PinotLeafPlanNode`, and change `groupByColumns` variable in it to `groupByExpressions`.

Added a task here: https://github.com/apache/pinot/issues/13957

In the quickstart I will add an example query leveraging this.